src/dbnode/storage/cleanup.go

```diff
@@ -301,26 +297,71 @@ func (m *cleanupManager) commitLogTimes(t time.Time) (time.Time, []time.Time, er
 	// are only retained for a period of 1-2 days (at most), after we which we'd live we with the
 	// data loss.
 
-	candidateTimes := timesInRange(earliest, latest, blockSize)
+	files, err := m.commitLogFilesFn(m.opts.CommitLogOptions())
+	if err != nil {
+		return nil, err
+	}
 	namespaces, err := m.database.GetOwnedNamespaces()
 	if err != nil {
-		return time.Time{}, nil, err
+		return nil, err
 	}
-	cleanupTimes := filterTimes(candidateTimes, func(t time.Time) bool {
+
+	var outerErr error
+	filesToCleanup := filterCommitLogFiles(files, func(start time.Time, duration time.Duration) bool {
+		if start.Before(earliest) {
+			// Safe to clean up expired files.
+			// TODO(rartoul): Now that we have commit log compaction via snapshot files we would like
+			// to remove the concept of commit log retention so that users cannot accidentally
+			// configure M3DB in such a way that it loses data (commit log retention < block size).
+			return true
+		}
+
 		for _, ns := range namespaces {
-			ropts := ns.Options().RetentionOptions()
-			start, end := commitLogNamespaceBlockTimes(t, blockSize, ropts)
-			if ns.NeedsFlush(start, end) {
+			var (
+				ropts                      = ns.Options().RetentionOptions()
+				nsBlocksStart, nsBlocksEnd = commitLogNamespaceBlockTimes(start, duration, ropts)
+				needsFlush                 = ns.NeedsFlush(nsBlocksStart, nsBlocksEnd)
+			)
+
+			if !needsFlush {
+				// Data has been flushed to disk so the commit log file is
+				// safe to clean up.
+				continue
+			}
+
+			// Add commit log blockSize to the startTime because that is the latest
+			// system time that the commit log file could contain data for. Note that
+			// this is different than the latest datapoint timestamp that the commit
+			// log file could contain data for (because of bufferPast/bufferFuture),
+			// but the commit log files and snapshot files both deal with system time.
+			isCapturedBySnapshot, err := ns.IsCapturedBySnapshot(
+				nsBlocksStart, nsBlocksEnd, start.Add(duration))
+			if err != nil {
+				outerErr = err
```

Comment:
If you want to get the first error btw, why not make it a slice `var outerErrs []error` then `outerErrs = append(outerErrs, err)` and then `if len(outerErrs) > 0 { return outerErrs[0] }`?

Reply:
Or even just use a `multiErr` here and return all errors *shrug*.